src/accountsdb/db.zig

```diff
@@ -2430,171 +2363,169 @@ pub const AccountsDB = struct {
-                .max_slot = target_slot,
+                .max_slot = params.target_slot,
             },
         });
 
         // TODO: this is a temporary value
         const delta_hash = Hash.default();
 
-        bank_fields.slot = target_slot; // !
-        bank_fields.capitalization = full_capitalization; // !
+        const archive_file = blk: {
+            const archive_file_name_bounded = sig.accounts_db.snapshots.FullSnapshotFileInfo.snapshotNameStr(.{
+                .slot = params.target_slot,
+                .hash = full_hash,
+                .compression = .zstd,
+            });
+            const archive_file_name = archive_file_name_bounded.constSlice();
+            break :blk try self.snapshot_dir.createFile(archive_file_name, .{ .read = true });
+        };
+        defer archive_file.close();
 
-        const package: FullSnapshotGenerationPackage = .{
-            .slot = target_slot,
+        const SerializableFileMap = std.AutoArrayHashMap(Slot, AccountFileInfo);
 
-            .lamports_per_signature = lamports_per_signature,
-            .bank_hash_info = .{
-                .accounts_delta_hash = delta_hash,
-                .accounts_hash = full_hash,
-                .stats = bank_hash_stats,
-            },
-            .bank_fields = bank_fields.*,
-            .deprecated_stored_meta_write_version = deprecated_stored_meta_write_version,
+        var serializable_file_map: SerializableFileMap, //
+        const bank_hash_stats //
+        = blk: {
+            var serializable_file_map = SerializableFileMap.init(self.allocator);
+            errdefer serializable_file_map.deinit();
+            try serializable_file_map.ensureTotalCapacity(file_map.count());
+
+            var bank_hash_stats = BankHashStats.zero_init;
+
+            for (file_map.values()) |account_file| {
+                if (account_file.slot > params.target_slot) continue;
 
-            .file_infos_map = serializable_file_map,
-            .file_map_rw = file_map_rw,
+                const bank_hash_stats_map, var bank_hash_stats_map_lg = self.bank_hash_stats.readWithLock();
+                defer bank_hash_stats_map_lg.unlock();
+
+                if (bank_hash_stats_map.get(account_file.slot)) |other_stats| {
+                    bank_hash_stats.accumulate(other_stats);
+                } else {
+                    self.logger.warnf("No bank hash stats for slot {}.", .{account_file.slot});
+                }
+
+                serializable_file_map.putAssumeCapacityNoClobber(account_file.slot, .{
+                    .id = account_file.id,
+                    .length = account_file.length,
+                });
+            }
+
+            break :blk .{ serializable_file_map, bank_hash_stats };
         };
-        const info: FullSnapshotGenerationInfo = .{
-            .slot = target_slot,
-            .hash = full_hash,
-            .capitalization = full_capitalization,
+        defer serializable_file_map.deinit();
```

Comment:
Mostly it's to declare `bank_hash_stats` as `const` thereafter. An alternative would be something like:
```zig
        var serializable_file_map = SerializableFileMap.init(self.allocator);
        defer serializable_file_map.deinit();

        const bank_hash_stats = blk: {
            var bank_hash_stats = BankHashStats.zero_init;

            try serializable_file_map.ensureTotalCapacity(file_map.count());
            for (file_map.values()) |account_file| {
                if (account_file.slot > params.target_slot) continue;

                const bank_hash_stats_map, var bank_hash_stats_map_lg = self.bank_hash_stats.readWithLock();
                defer bank_hash_stats_map_lg.unlock();

                if (bank_hash_stats_map.get(account_file.slot)) |other_stats| {
                    bank_hash_stats.accumulate(other_stats);
                } else {
                    self.logger.warnf("No bank hash stats for slot {}.", .{account_file.slot});
                }

                serializable_file_map.putAssumeCapacityNoClobber(account_file.slot, .{
                    .id = account_file.id,
                    .length = account_file.length,
                });
            }

            break :blk bank_hash_stats;
        };
```
But that looks slightly awkward since it makes it look like at a glance that the declaration of `bank_hash_stats` and `serializable_file_map` are disconnected, despite being tied together.

The version where we eschew the block kind of has the worst of both worlds in that respect, where it makes `bank_hash_stats` unnecessarily mutable after initialization, and makes their initialization blend in with the rest of the code, hiding their relationship:
```zig
        var serializable_file_map = SerializableFileMap.init(self.allocator);
        defer serializable_file_map.deinit();

        var bank_hash_stats = BankHashStats.zero_init;

        try serializable_file_map.ensureTotalCapacity(file_map.count());
        for (file_map.values()) |account_file| {
            if (account_file.slot > params.target_slot) continue;

            const bank_hash_stats_map, var bank_hash_stats_map_lg = self.bank_hash_stats.readWithLock();
            defer bank_hash_stats_map_lg.unlock();

            if (bank_hash_stats_map.get(account_file.slot)) |other_stats| {
                bank_hash_stats.accumulate(other_stats);
            } else {
                self.logger.warnf("No bank hash stats for slot {}.", .{account_file.slot});
            }

            serializable_file_map.putAssumeCapacityNoClobber(account_file.slot, .{
                .id = account_file.id,
                .length = account_file.length,
            });
        }
```
We can vaguely represent this through comments, but imo simply writing these things as part of the code structure is more clear and leaves less room for any doubt, and retains the nice property that the reader can be sure that `bank_hash_stats` isn't modified thereafter (ideally `serializable_file_map` would also be `const` to signify the same, but the deinit method is poorly written and takes a mutable reference).

Reply:
I think comments are an extremely subpar tool for representing such things, being subject to desync with the actual code and all, but I'll concede for the sake of moving forward.

Reply:
f77f5ec06afd88e094da647cc5177e299093e362

Reply:
thank you 🙇